File: src/test/java/racingcar/domain/CarTest.java
package racingcar.domain;

import static org.assertj.core.api.Assertions.*;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullAndEmptySource;
import org.junit.jupiter.params.provider.ValueSource;

class CarTest {

	@ParameterizedTest
	@DisplayName("자동차 이름은 null 값을 가지거나 비어있을 수 없다.")
	@NullAndEmptySource
	@ValueSource(strings = {" ", "  "})
	void constructorWithEmptyCarNames(String carName) {
		assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() -> new Car(carName));
	}

	@ParameterizedTest
	@DisplayName("자동차 이름은 5글자를 넘을 수 없다.")
	@ValueSource(strings = {"동해물과백두", "동해물과백두산"})
	void constructorWithCarNamesExceedMaximumLength(String carName) {
		assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() -> new Car(carName));
	}

	@ParameterizedTest
	@DisplayName("자동차는 5글자 이하의 이름을 가질 수 있다.")
	@ValueSource(strings = {"동", "동해", "동해물", "동해물과", "동해물과백"})
	void constructor(String carName) {
		assertThatCode(() -> new Car(carName)).doesNotThrowAnyException();
	}

	@Test
	@DisplayName("숫자가 4 이상인 경우 자동차는 전진한다.")
	void move() {
		final var car = new Car("홍길동");
		car.move(new GoForwardStrategy());

		assertThat(car.position()).isEqualTo(1);
	}

	@Test
	@DisplayName("숫자가 4 미만인 경우 자동차는 전진하지 못한다.")
	void stop() {
		final var car = new Car("홍길동");
		car.move(new StopStrategy());

		assertThat(car.position()).isEqualTo(0);
	}
}

File: src/test/java/racingcar/domain/CarTest.kt
package racingcar.domain

import io.kotest.assertions.throwables.shouldNotThrow
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FreeSpec
import io.kotest.matchers.shouldBe
import racingcar.strategy.move.RandomMoveStrategy
import java.lang.IllegalArgumentException

class CarTest : FreeSpec({

    "자동차의 이름의 길이는 0 이하일 수 없다" {
        shouldThrow<IllegalArgumentException> {
            Car(Name(""))
        }.run {
            message shouldBe "이름은 빈 값일 수 없습니다."
        }
    }

    "자동차의 이름의 길이는 6 이상일 수 없다" {
        shouldThrow<IllegalArgumentException> {
            Car(Name("a".repeat(6)))
        }.run {
            message shouldBe "이름의 길이는 1 ~ 5 사이여야 합니다."
        }
    }

    "자동차의 이름의 길이는 1 ~ 5 글자여야 한다" - {
        for (i in 1..5) {
            "글자 길이가 $i 인 경우" - {
                shouldNotThrow<IllegalArgumentException> {
                    Car(Name("a".repeat(i)))
                }
            }
        }
    }

    "자동차의 위치는 0 부터 시작된다" {
        Car(name = Name("현대")).run {
            position shouldBe Position(0)
        }
    }

    "자동차는 숫자 값 4 이상이 주어진 경우 한 칸 움직일 수 있다" {
        Car(Name("현대")).moveForward(RandomMoveStrategy({ 4 })).run {
            position shouldBe Position(1)
        }
    }

    "자동차는 숫자 값 4 미만이 주어진 경우 움직일 수 없다" {
        Car(Name("현대")).moveForward(RandomMoveStrategy({ 3 })).run {
            position shouldBe Position(0)
        }
    }
})
